### PR TITLE
Feature/improve normalization pseudonym generator

### PR DESCRIPTION
**Pull request by izcram**
* right aligning variants, in particular also intronic variants, wrt to cDNA (i.e. right or left align on genomic coordinates, depending on the orientation/strand of the gene)
* more exhaustive synonyms generation, also including left aligned variants
* parallelizing many computations in the brca pseudonym generator using `pathos` (requires a rebuild of the docker image)

### CODE REVIEW

File: pipeline/data_merging/brca_pseudonym_generator.py [hunk 1]
```diff
@@ -165,89 +168,153 @@ def get_synonyms(x, hgvs_proc, syn_ac_dict):
...
-                    if vp_norm:
-                        if vp_norm not in synonyms:
-                            logging.info("Found new synonym! " + str(vp_norm) + " " + str(vp) + " " + str(x[PYHGVS_GENOMIC_COORDINATE_38_COL]))
-                        synonyms.append(vp_norm)
                 except HGVSError as e:
-                    logging.info("Exception in synonym handling " + str(vc) + " from " + str(vc.ac) + " to " +
-                                 str(dst) + " using " + str(method) + " via " + str(alt_ac) + " : " + str(e) + " " + str(e.__class__.__name__))
+                    logging.info("Erowception in synonym handling " + str(vc) + " from " + str(vc.ac) + " to " +
+                                 str(dst) + " using " + str(method) + " via " + str(alt_ac) + " : " + str(
+                        e) + " " + str(e.__class__.__name__))
```
Comment (zfisch) [inline pipeline/data_merging/brca_pseudonym_generator.py:174]: typo?

File: pipeline/data_merging/brca_pseudonym_generator.py [hunk 2]
```diff
@@ -165,89 +168,153 @@ def get_synonyms(x, hgvs_proc, syn_ac_dict):
...
+
+    logging.info("Converting variants to hgvs objects")
+
+    df[TMP_HGVS_HG38] = df[VAR_OBJ_FIELD].apply(
+        lambda v: v.to_hgvs_obj(hgvs_proc.contig_maps[HgvsWrapper.GRCh38_Assem]))
+
+    logging.info("Normalize genomic representation")
+
+    df = utils.parallelize_dataframe(df, _normalize_genomic_fnc(TMP_HGVS_HG38, TMP_HGVS_HG38, True, strand_dict), processes)
+    df = utils.parallelize_dataframe(df, _normalize_genomic_fnc(TMP_HGVS_HG38, TMP_HGVS_HG38_LEFT_ALIGNED, False, strand_dict),
```
Comment (zfisch) [inline pipeline/data_merging/brca_pseudonym_generator.py:225]: i found this a bit confusing, could use a comment to explain why normalize_genomic_fnc needs to be run twice like this

File: pipeline/data_merging/brca_pseudonym_generator.py [hunk 3]
```diff
@@ -165,89 +168,153 @@ def get_synonyms(x, hgvs_proc, syn_ac_dict):
...
+    # normalizing again for the hg37 representation. An alternative would be to convert the normalized hg38 representation to hg37.
+    # If we use crossmap, we would need a way to convert the VCF like representation back to an hgvs object, which we currently
+    # are unable to do properly. That is, we can use VCFVariant.to_hgvs_obj, however, structural variants will be converted
+    # to delins, losing information if a variant was e.g. a del, ins, or dup.
+
+    df[TMP_HGVS_HG37] = pd.Series(
+        [v.to_hgvs_obj(hgvs_proc.contig_maps[HgvsWrapper.GRCh37_Assem]) for v in var_objs_hg37])
+
+    df = utils.parallelize_dataframe(df, _normalize_genomic_fnc(TMP_HGVS_HG37, GENOMIC_HGVS_HG37_COL, True, strand_dict), processes)
+    df[GENOMIC_HGVS_HG37_COL] = df[GENOMIC_HGVS_HG37_COL].apply(str)
```
Comment (zfisch) [inline pipeline/data_merging/brca_pseudonym_generator.py:243]: same as above -- how come it takes two immediate runs of _normalize_genomic_fnc to get desired output?

**Review by zfisch — COMMENTED** [review]
awesome, i'm attempting to run it and will report back
left a couple questions regarding _normalize_genomic_fnc and why it's always run twice in quick succession (would it make sense to update it to do the same thing in a single run?)

File: pipeline/data_merging/brca_pseudonym_generator.py (hunk 1 above)
Comment (izcram) [reply pipeline/data_merging/brca_pseudonym_generator.py:174]: yes, indeed! (replace all gone wrong)

File: pipeline/data_merging/brca_pseudonym_generator.py (hunk 2 above)
Comment (izcram) [reply pipeline/data_merging/brca_pseudonym_generator.py]: Once we align right wrt to cDNA and once we align left (for the synonyms)

**Review by zfisch — APPROVED** [review]
test run was successful 👍

**Pull request merged**